src/buffers/disk/leveldb_buffer.rs

```diff
@@ -171,15 +171,21 @@ impl Stream for Reader {
         // using write_notifier to wake this task up after the next write.
         self.write_notifier.register();
 
-        // This will usually complete instantly, but in the case of a large queue (or a fresh launch of
-        // the app), this will have to go to disk.
-        let next = tokio::task::block_in_place(|| {
-            self.db
-                .get(ReadOptions::new(), Key(self.read_offset))
-                .unwrap()
-        });
+        if self.buffer.is_empty() {
+            // This will usually complete instantly, but in the case of a large queue (or a fresh launch of
+            // the app), this will have to go to disk.
+            let new_data = tokio::task::block_in_place(|| {
+                self.db
+                    .value_iter(ReadOptions::new())
+                    .from(&Key(self.read_offset))
+                    .to(&Key(self.read_offset + 100))
+                    .collect()
+            });
+            self.buffer = new_data;
+            self.buffer.reverse(); // so we can pop
```

Comment:
I think we could probably speed this up and avoid the reverse. Vec implements doubleended iterator so you can next_back.

Also maybe interesting to just work directly with https://docs.rs/leveldb/0.8.4/leveldb/database/iterator/struct.ValueIterator.html and seek to the last element and go back from there. That might be faster?

Reply:
I left this pretty naive because I'm skeptical that reversing a hundred-element vec is terribly meaningful perf-wise in this case. But that's certainly something we could try and benchmark :smile:

Reply:
I'm sure one vec reversal would give us at least another 50% 😄 

Joking aside, I think there is potential to extract a bit more by using the actual leveldb iterator but who knows, we can follow up later.